src/cdk/v2/destinations/the_trade_desk/transformRecord.js

```diff
@@ -0,0 +1,90 @@
+const { InstrumentationError, ConfigurationError } = require('@rudderstack/integrations-lib');
+const { BatchUtils } = require('@rudderstack/workflow-engine');
+const {
+  defaultPostRequestConfig,
+  defaultRequestConfig,
+  getSuccessRespEvents,
+  removeUndefinedAndNullValues,
+  handleRtTfSingleEventError,
+} = require('../../../../v0/util');
+const { getTTLInMin, getFirstPartyEndpoint } = require('./utils');
+const tradeDeskConfig = require('./config');
+
+const { DATA_PROVIDER_ID } = tradeDeskConfig;
+
+const responseBuilder = (items, config) => {
+  const { advertiserId, dataServer } = config;
+
+  const payload = { DataProviderId: DATA_PROVIDER_ID, AdvertiserId: advertiserId, Items: items };
+
+  const response = defaultRequestConfig();
+  response.endpoint = getFirstPartyEndpoint(dataServer);
+  response.method = defaultPostRequestConfig.requestMethod;
+  response.body.JSON = removeUndefinedAndNullValues(payload);
```

Comment:
For warehouse as sources, null is a valid column value --> so that should be propagated to the destination. Otherwise, any changes in warehouse table on updating field value from not-null to null will never propagate to the destination leading to inconsistency.
Can you verify this & see if there are any issues from TTD end if null values passed ?

Reply:
Trade desk don't accept the null audience Ids.
It throw an error like this

> {"FailedLines":[{"ErrorCode":"MissingUserId","Message":"Missing TDID DAID UID2 IDL CoreID EUID ID5 NetID and FirstID, item #1"}]}

Reply:
This should be for mandatory fields. In 1P flow, don't we have any optional attributes in payload ?

Reply:
Only ID is coming from warehouse columns, that is mandatory.
Other optional properties are coming from webapp input